### PR TITLE
Implement get_semantic_zones() for ClientPane

### LINKED ISSUE #5141: Jump to semantic regions not working through non-local domains
### What Operating System(s) are you seeing this problem on?

Linux X11

### Which Wayland compositor or X11 Window manager(s) are you using?

_No response_

### WezTerm version

wezterm 20240102-011725-2f7ca41a

### Did you try the latest nightly build to see if the issue is better (or worse!) than your current version?

Yes, and I updated the version box above to show the version of the nightly that I tried

### Describe the bug

In non-local domains (i.e., ssh or mux domains, at least), jumping with semantic regions is not working. The jumps simply do nothing.

### To Reproduce

1. Enable semantic region in shell and wezterm
2. Start mux domain with e.g. unix domain
3. Goto copy mode and jump

### Configuration

Since by default moving by zone type is not accessible from the command palette: 
```lua
local wezterm = require 'wezterm'
local act = wezterm.action
return {
    key_tables = {
        copy_mode = {
            { key = '(', mods = 'SHIFT', action = act.CopyMode { MoveBackwardZoneOfType = 'Prompt' } },
            { key = ')', mods = 'SHIFT', action = act.CopyMode { MoveForwardZoneOfType = 'Prompt' } },
        }
    }
}
```

### Expected Behavior

_No response_

### Logs

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by Osse**
By adding a new request and response we can query the remote terminal
for its semantic zones. Thus handy features like the ScrollToPrompt and
CopyMode actions work as expected in other domains than the local
domain.

Fixes #5141

---------------------------------------------------

I implemented this fairly quickly buy just basing it on the patterns in
surrouding code, and it seems to work exactly as I had hoped. Some
concerns/points:

- As far as I can tell the RPC stuff needs to be `async` functions while
  `get_semantic_zones()` from the `Pane` trait is not. I worked around this by
  using `smol::block_on()`. I see that is used many places in the code base but I'm
  not sure it's acceptable here. An alternative is to make the function `async`
  in the trait, but that seemed messy for other reasons.
- I couldn't find any places to add a test for this.
- There is no new documentation. Imo. that's justified as I am making an
  existing feature work in more contexts which the existing documentation did
  not clarify that it _didn't_ work.

### CODE REVIEW

File: codec/src/lib.rs [hunk 1]
```diff
@@ -502,6 +502,8 @@ pdu! {
     GetPaneDirection: 60,
     GetPaneDirectionResponse: 61,
     AdjustPaneSize: 62,
+    SemanticZonesRequest: 63,
```
Comment (wez) [inline codec/src/lib.rs:506]: Please increment `CODEC_VERSION` (on line 444 above this) when adding new PDUs; that will enable the client to realize when the server doesn't support the PDU and error out and inform the user.

File: wezterm-client/src/pane/clientpane.rs [hunk 2]
```diff
@@ -425,6 +425,20 @@ impl Pane for ClientPane {
         Ok(())
     }
 
+    fn get_semantic_zones(&self) -> anyhow::Result<Vec<SemanticZone>> {
+        match smol::block_on(async {
```
Comment (wez) [inline wezterm-client/src/pane/clientpane.rs]: we must not use block_on here, as it will block the GUI thread that may be calling into this code, and can lead to deadlock in various scenarios

The proper fix would be to change the signature of `get_semantic_zones` in the trait definition in `mux/src/pane.rs` to be async, then to address the fan-out from that to add appropriate awaits in the all the places that the compiler will identify

**Review by wez — CHANGES_REQUESTED** [review]
Thanks for diving in!
I think the approach looks fine, but we have to implement this without using `block_on`.
I think you can probably get quite far (maybe even 100%) with making this async just by changing the signature of the trait method and seeing what the compiler complains about; likely just then adding some `.awaits`.

Could you give that a try and update the PR?
Thanks!

File: codec/src/lib.rs (hunk 1 above)
Comment (Osse) [reply codec/src/lib.rs:506]: Done!

File: wezterm-client/src/pane/clientpane.rs (hunk 2 above)
Comment (Osse) [reply wezterm-client/src/pane/clientpane.rs]: I suspected this was the case. I pushed 0cdf12ab5 to make it async, but I ended up adding even more `smol::block_on()` because of it. It's WIP. Hopefully I can spend some more time on it today.

Comment (Osse) [reply wezterm-client/src/pane/clientpane.rs]: Made some progress and have pushed 2452d2764, replacing the previous one. I assume this will all be squashed in the end (either by you or me). This just shows the changes needed from my first proposal to making it async.

There are three `smol::block_on()` calls left that I don't know what to do with. Looking at the nearby code `promise::spawn::spawn(...).detach()` is used a lot. It seems  to be used in cases where you just "fire and forget" something, which doesn't work in this case I believe.

**Comment by Osse** [comment]
I have been using Wezterm with this patch for a few months now and everything seems to work very well. I have not noticed any regressions, locally nor over SSH. I took it upon myself to rebase and squash the changes as I mentioned above.

I haven't had much time to improve the patch, though, and there are still a few `block_on()` that I don't know how to get rid of.
